docs/specs/instrumentation.md

```diff
@@ -8,34 +8,103 @@ interpreted as described in [RFC 2119].
 
 ## Naming
 
+### Event naming
+
+Vector implements an event-driven instrumentation pattern ([RFC 2064]) and
+event names MUST adhere to the following rules:
+
+* MUST only contain ASCII alphanumeric and lowercase characters
+* MUST be in [camelcase] format
+* MUST follow the `<Namespace><Noun><Verb>[Error]` template
+  * `Namespace` - the internal domain the event belongs to (e.g., `Component`, `Buffer`, `Topology`)
+  * `Noun` - the subject of the event (e.g., `Bytes`, `Events`)
+  * `Verb` - the past tense verb describing when the event occured (e.g., `Received`, `Sent`, `Processes`)
+  * `[Error]` - if the event is an error it MUST end with `Error`
+
 ### Metric naming
 
-For metric naming, Vector broadly follows the
-[Prometheus metric naming standards]. Hence, a metric name:
+Vector broadly follows the [Prometheus metric naming standards]:
 
-* MUST only contain ASCII alphanumeric, lowercase, and underscores
+* MUST only contain ASCII alphanumeric, lowercase, and underscore characters
+* MUST be in [snakecase] format
 * MUST follow the `<namespace>_<name>_<unit>_[total]` template
-  * `namespace` represents a broad category of metrics (e.g., `component`, `buffer`, `topology`)
-  * `name` is one or more words that describes the measurement (e.g., `memory_rss`, `requests`)
-  * `unit` MUST be a single [base unit] in plural form, if applicable (e.g., `seconds`, `bytes`)
+  * `namespace` - the internal domain that the metric belongs to (e.g., `component`, `buffer`, `topology`)
+  * `name` - is one or more words that describes the measurement (e.g., `memory_rss`, `requests`)
+  * `unit` - MUST be a single [base unit] in plural form, if applicable (e.g., `seconds`, `bytes`)
   * Counters MUST end with `total` (e.g., `disk_written_bytes_total`, `http_requests_total`)
 * SHOULD be broad in purpose and use use tags to differentiate characteristics of the measurement (e.g., `host_cpu_seconds_total{cpu="0",mode="idle"}`)
 
+## Emission
+
+### Batching
+
+For performance reasons, as demonstrated in ppull request #8383],
+instrumentation SHOULD be batched whenever possible:
+
+* Vector process batches of events ([RFC 9480]) and telemtry SHOULD emit for
+  the entire batch, not each individual event.
+
+### Errors
+
+As described in the [events](#events) section, all errors must emit as events
+to drive log and metric emission. Because errors can be transient and
+recoverable, errors MUST be able to log at warning and error levels:
+
+* MUST emit at the [error level](#error-level) if the error requires user
+  attention, otherwise the error MUST emit at the [warning level](#warning-level)
+  * Retryable errors only require user attention when the retry count is >= 3
+
+#### Warning level
+
+When an error event emits at the warning level it does not require user
+attention and, therefore, MUST do the following:
+
+* MUST log a message at the `warning` level
+* MUST NOT increment the `component_errors_total`
```

Comment:
Reading this now, I'm not crazy about component specific metric information being in the `intrumentation.md` spec. Maybe we could separate these bits specific to components back out into the `component.md` spec?

Reply:
Yeah, I can generalize this requirement. I didn't move anything component specific into the instrumentation doc.